### ISSUE TITLE
[BUG] Investigate odd AST slowness with ASYNC allocator.

### ISSUE THREAD

**Issue by revans2** [NVIDIA/spark-rapids#4536]
**Describe the bug**
I am having some really odd problems with the ASYNC allocator. So it could just be related to my setup specifically. I have been testing the performance of AST against binops in a number of situations. I found that when using the ASYNC allocator and a large number of rows (100000000) many of the AST runs get really slow. Like on the order of seconds slow.  I have seen this happen with BOOL8 doing (LOGICAL_AND, LOGICAL_OR) and (FLOAT32, FLOAT64, INT32, INT64) doing (ADD, DIV, MUL, SUB, EQUAL, GREATER, GREATER_EQUAL, LESS, LESS_EQUAL, and NOT_EQUAL)

It is especially odd because it does not happen if there are nulls in the data.  It does not happen if AST runs before binops, only after. It does not happen with smaller amounts of data. And for some of the operators it only happens when there are a small number of operations being done. It does not happen with no polling, and it does not happen with ARENA.

I have also seen the test hang on shutdown, and the only way to stop it is with a kill -9, but that may be a separate issue.

**Steps/Code to reproduce bug**
[DataTest.scala.zip](https://github.com/NVIDIA/spark-rapids/files/7872187/DataTest.scala.zip)

**Expected behavior**
AST is not 500 times slower in this odd corner case.

**Comment by rongou** [comment]
Was able to reproduce this on my desktop. The profile looks pretty funky:
```console
CUDA API Statistics:

 Time (%)  Total Time (ns)  Num Calls    Avg (ns)       Med (ns)      Min (ns)      Max (ns)      StdDev (ns)                  Name                
 --------  ---------------  ---------  -------------  -------------  -----------  -------------  -------------  -----------------------------------
     49.3    1,722,493,522         10  172,249,352.2        1,302.5          871  1,722,475,770  544,694,041.0  cudaStreamSynchronize_ptsz         
     27.9      975,786,305      1,605      607,966.5        4,288.0        3,737      2,062,738      817,300.2  cudaLaunchKernel_ptsz              
     16.9      590,838,738          1  590,838,738.0  590,838,738.0  590,838,738    590,838,738            0.0  cudaFree                           
      5.1      177,420,710          1  177,420,710.0  177,420,710.0  177,420,710    177,420,710            0.0  cudaMemPoolDestroy_v11020          
      0.8       28,256,624      1,610       17,550.7          802.0          721     26,893,473      670,224.6  cudaMallocFromPoolAsync_ptsz_v11020
      0.0        1,517,895      1,610          942.8          902.0          761          6,883          245.8  cudaFreeAsync_ptsz_v11020          
      0.0          524,311         11       47,664.6       10,239.0        8,546        384,798      111,974.4  cudaMemcpyAsync_ptsz               
      0.0           37,030          1       37,030.0       37,030.0       37,030         37,030            0.0  cudaMemGetInfo                     
      0.0           35,919          7        5,131.3        2,706.0          351         23,825        8,338.6  cudaMemsetAsync_ptsz               
      0.0           22,212          1       22,212.0       22,212.0       22,212         22,212            0.0  cudaMemPoolCreate_v11020           
      0.0            1,423          1        1,423.0        1,423.0        1,423          1,423            0.0  cudaMemPoolSetAttribute_v11020     
```

compared to using ARENA:
```console
CUDA API Statistics:

 Time (%)  Total Time (ns)  Num Calls    Avg (ns)       Med (ns)     Min (ns)    Max (ns)     StdDev (ns)              Name           
 --------  ---------------  ---------  -------------  -------------  ---------  -----------  -------------  --------------------------
     76.2    1,971,947,332      1,950    1,011,255.0    1,665,288.5        801    2,191,948      968,938.8  cudaStreamSynchronize_ptsz
     23.3      603,061,085          2  301,530,542.5  301,530,542.5  4,756,422  598,304,663  419,701,986.2  cudaFree                  
      0.2        4,951,857      1,030        4,807.6        4,449.0      3,647       26,110        1,532.0  cudaLaunchKernel_ptsz     
      0.2        4,633,589          1    4,633,589.0    4,633,589.0  4,633,589    4,633,589            0.0  cudaMalloc                
      0.1        3,650,695        921        3,963.8        3,306.0      2,826      374,891       12,311.3  cudaMemcpyAsync_ptsz      
      0.0          189,127        462          409.4          281.0        250       42,721        1,983.8  cudaMemsetAsync_ptsz      
      0.0           30,237          1       30,237.0       30,237.0     30,237       30,237            0.0  cudaMemGetInfo          
```

**Comment by rongou** [comment]
Ok, I think I figured this out. The slowness is caused by the asyn nature of the ASYNC allocator. Since we are running binops and asts back to back, binops accumulates a bunch of cuda operations in the stream, leading to the asts to slow down synchronizing the stream. If we add a stream synchronize call:
```scala
...
val (numBinIters, binTotalTime) = runUntil(maxTestIterations, targetTestTimeNs,
  binops(t, regOp, DType.BOOL8, regOp, DType.BOOL8).close())
Cuda.DEFAULT_STREAM.sync()  
val (numAstIters, astTotalTime) = runUntil(maxTestIterations, targetTestTimeNs,
  asts(t, astOp, astOp).close())
...
```
then the result is normal:
```console
BOOL8	LOGICAL_AND	1	100000000	false	486	2.058715024691358	90.47610411791715	1609	0.6225212392790553	299.20989545482735
```
